### PR TITLE
Add 1st draft ofn et

### PR DESCRIPTION
**Pull request by Rulison**

### CODE REVIEW

File: dual_net.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+def residual_block(tensor, specs):
+    input_tensor = tensor
+    tensor = conv_block(tensor, specs)
+    tensor = layers.convolution2d(tensor,
+                                  num_outputs=specs['num_outputs'],
+                                  kernel_size=specs['kernel_size'],
+                                  stride=specs['stride'],
+                                  activation_fn=None)
+    tensor = tf.layers.batch_normalization(tensor)
+    tensor += input_tensor
```
Comment (thomas-a-neil) [inline dual_net.py:27]: So this step will only work if stride=1? Otherwise the output will be a different size than the input?

File: dual_net.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    tensor = tf.nn.relu(tensor)
+    return tensor
+
+
+def build_mlp(board_placeholder,
```
Comment (thomas-a-neil) [inline dual_net.py]: Just curious, why did you call this `build_mlp`? I thought "mlp" referred to specifically a NN made up of fully connected layers.

File: dual_net.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+       - d['layer'] <- 'fc'
+       - d['num_outputs'] is an int
+       - d['activation_fn'] is a function
+    policy_head and value_head have the same structure as above but represent
+    the layers for the policy head and value head, respectively.
+
+    returns the policy output and the value output in a tuple
+    """
+    out = board_placeholder
+    for specs in shared_layers:
```
Comment (thomas-a-neil) [inline dual_net.py:58]: So this loop only works if the conv blocks come first, followed by residual layers, right? Because the residual layers execute a flattening of the tensor, so you couldn't do a convolution afterwards.

I'm pointing this out because it seems this loop isn't really robust to different inputs... I wonder if we should just hardcode some of this instead of pretending that we can pass in any dictionary we want.

File: dual_net.py (hunk 2 above)
Comment (Rulison) [reply dual_net.py]: No reason other than that's what it was called in the homeworks lol

File: dual_net.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    def __init__(self, sess, action_rep='pos', learning_rate=0.01, regularization_mult=0.01):
+        """
+        sess: tensorflow session
+        action_rep: 'pos' for (pos1, pos2) action representation, 'piece' for (piece, pos) rep
+        learning_rate: learning rate for gradient descent
+        regularization_mult: multiplier for weight regularization loss
+        """
+        self.board_placeholder = tf.placeholder(tf.float32, [None, 8, 8, 13])
+        shared_layers = [{'layer': 'conv', 'num_outputs': 256, 'stride': 3, 'kernel_size': 1, 'activation_fn': tf.nn.relu},
+                         {'layer': 'residual', 'num_outputs': 256, 'stride': 1, 'kernel_size': 3, 'activation_fn': tf.nn.relu},
```
Comment (thomas-a-neil) [inline dual_net.py]: To clarify, they have 20 or 40 residual layers right? Are we just starting with two for now?

File: dual_net.py (hunk 3 above)
Comment (Rulison) [reply dual_net.py:58]: Correct, as in the paper, which is one conv layer followed by a number of residual layers.  I can hardcode that.

File: dual_net.py (hunk 4 above)
Comment (Rulison) [reply dual_net.py]: Yeah Josh recommended starting with a smaller net to see how well that performs.

File: dual_net.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        value_layers = [{'layer': 'conv', 'num_outputs': 1, 'stride': 1, 'kernel_size': 1, 'activation_fn': tf.nn.relu},
+                        {'layer': 'fc', 'num_outputs': 256, 'activation_fn': tf.nn.relu},
+                        {'layer': 'fc', 'num_outputs': 1, 'activation_fn': tf.nn.tanh}]
+
+        self.policy_predict, self.value_predict = build_mlp(self.board_placeholder,
+                                                            scope='net',
+                                                            shared_layers=shared_layers,
+                                                            policy_head=policy_layers,
+                                                            value_head=value_layers)
+        self.z = tf.placeholder(tf.float32, [None])
```
Comment (thomas-a-neil) [inline dual_net.py]: If we're going to call the prediction `value_predict` maybe we can call this `value_observed`?

File: tests/test_net.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        self.boards = np.random.random_sample([10, 8, 8, 13])
+        self.z = np.random.random_sample([10])
+
+    def test_predict(self):
+        policy, value = self.net.predict(self.boards)
+        self.assertEqual(policy.shape, (10, 64*64))
+
+    def test_train(self):
+        pi = np.random.random_sample([10, 64*64])
+        # violates unit testing protocol, but w/e
```
Comment (thomas-a-neil) [inline tests/test_net.py]: So sometimes these tests pass and sometimes they fail? '

I wonder if it's worth just doing one run, computing `self.net.loss, self.net.regularization_loss` and a prediction, and then just asserting that their shapes are all as expected... 

Do you think that asserting that the loss goes down will be something we'll want to test frequently?

File: dual_net.py (hunk 4 above)
Comment (thomas-a-neil) [reply dual_net.py]: Nice yeah I agree! I say we add that bad boy in as a parameter - I'll push something with it

File: dual_net.py (hunk 2 above)
Comment (thomas-a-neil) [reply dual_net.py]: I changed it to "build_model" which I think is more honest 😛

File: tests/test_net.py [hunk 7]
```diff
@@ -21,25 +21,9 @@ def test_predict(self):
         policy, value = self.net(self.boards)
         self.assertEqual(policy.shape, (10, 64*64))
 
-    def test_train(self):
```
Comment (thomas-a-neil) [inline tests/test_net.py]: Nice I'm in favor of removing this
